### PR TITLE
[shardformer] support 'paddedcausal'  type of attention mask in Coloattention.

### PR DESCRIPTION
**Pull request by flybird1111**
## 📌 Checklist before creating the PR

- [x] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4184 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

support 'paddedcausal' of attention mask type in Coloattention.

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/kernel/cuda_native/flash_attention.py [hunk 1]
```diff
@@ -86,11 +86,11 @@ def backward(ctx, grad_output):
 
     class ColoAttention(torch.nn.Module):
 
-        def __init__(self, embed_dim: int, num_heads: int, dropout: float = 0.0):
+        def __init__(self, embed_dim: int, num_heads: int, dropout: float = 0.0, scale=1.0):
             super().__init__()
             assert embed_dim % num_heads == 0, \
                 f"the embed dim ({embed_dim}) is not divisible by the number of attention heads ({num_heads})."
-            self.scale = 1 / math.sqrt(embed_dim // num_heads)
+            self.scale = scale
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/flash_attention.py:93]: How about check if the input scale is None, we use 1 / sqrt(embed_dim / num_heads), else we set its value

Comment (flybird1111) [reply colossalai/kernel/cuda_native/flash_attention.py]: > How about check if the input scale is None, we use `1 / sqrt(embed_dim / num_heads)`, otherwise we set its value

updated

File: colossalai/kernel/cuda_native/flash_attention.py [hunk 2]
```diff
@@ -146,7 +157,7 @@ def forward(self,
 
             out = memory_efficient_attention(query, key, value, attn_bias=attn_bias, p=self.dropout, scale=self.scale)
 
-            if attn_mask_type == AttnMaskType.padding and batch_size > 1:
+            if attn_mask_type in [AttnMaskType.padding, AttnMaskType.paddedcausal] and batch_size > 1:
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/flash_attention.py:160]: How about this way to check if padded: `attn_mask_type.value % 2 == 1`

File: colossalai/kernel/cuda_native/scaled_softmax.py [hunk 3]
```diff
@@ -151,7 +152,7 @@ def forward_fused_softmax(self, input, mask):
         b, np, sq, sk = input.size()
         scale = self.scale if self.scale is not None else 1.0
 
-        if self.attn_mask_type == AttnMaskType.causal:
+        if self.attn_mask_type in [AttnMaskType.causal, AttnMaskType.paddedcausal]:
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/scaled_softmax.py:155]: Similarly, check if causal by `attn_mask_type.value > 1`

File: colossalai/kernel/cuda_native/flash_attention.py (hunk 2 above)
Comment (flybird1111) [reply colossalai/kernel/cuda_native/flash_attention.py]: > How about this way to check if padded: `attn_mask_type.value % 2 == 1`

Updated.

**Review by kurisusnowdeng — APPROVED** [review]
